File: 0505-the-maze-ii/0505-the-maze-ii.py
class Solution:
    def dijkstra(self, maze: List[List[int]], start: List[int], distances: List[List[int]]) -> None:
        pq = [(0, start[0], start[1])]
        while pq:
            curr, x, y = heappop(pq)
            if distances[x][y] < curr:
                continue
            
            for dx, dy in (1, 0), (-1, 0), (0, 1), (0, -1):
                new_x, new_y, dist = x + dx, y + dy, 0
                while 0 <= new_x < self.m and 0 <= new_y < self.n and maze[new_x][new_y] == 0:
                    new_x += dx
                    new_y += dy
                    dist += 1
                if distances[x][y] + dist < distances[new_x - dx][new_y - dy]:
                    distances[new_x - dx][new_y - dy] = distances[x][y] + dist
                    heappush(pq, (distances[new_x - dx][new_y - dy], new_x - dx, new_y - dy))                
    
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        self.m, self.n = len(maze), len(maze[0])
        distances = [[float('inf')] * self.n for _ in range(self.m)]
        distances[start[0]][start[1]] = 0
        
        self.dijkstra(maze, start, distances)
        return -1 if distances[destination[0]][destination[1]] == float('inf') else distances[destination[0]][destination[1]]

File: 0505-the-maze-ii/0505-the-maze-ii.cpp
class Solution {
private:
    int m, n;
    bool isValid(int x, int y, vector<vector<int>>& maze){
        return 0 <= x && x < m && 0 <= y && y < n && maze[x][y] == 0;
    }
    
public:
    int shortestDistance(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        this -> m = maze.size(), this -> n = maze[0].size();
        queue<pair<int, int>> queue;
        queue.push(make_pair(start[0], start[1]));
        vector<vector<int>> distances(m, vector<int>(n, INT_MAX));
        distances[start[0]][start[1]] = 0;
        
        while (!queue.empty()){
            auto [x, y] = queue.front();
            queue.pop();
            
            for (auto& [dx, dy] : vector<pair<int, int>>{{1, 0}, {-1, 0}, {0, -1}, {0, 1}}){
                int newX = x + dx, newY = y + dy, dist = 0;
                
                while (isValid(newX, newY, maze)){
                    newX += dx;
                    newY += dy;
                    ++dist;
                }
                
                if (distances[x][y] + dist < distances[newX - dx][newY - dy]){
                    distances[newX - dx][newY - dy] = distances[x][y] + dist;
                    queue.push({newX - dx, newY - dy});
                }
            }
        }
        
        return distances[destination[0]][destination[1]] == INT_MAX? -1 : distances[destination[0]][destination[1]];
    }
};